update/v1alpha1/types_update_status_progress.go

```diff
@@ -0,0 +1,424 @@
+	// Progressing means the control plane is updating and no problems or slowness are detected
+	ControlPlaneAssessmentProgressing ControlPlaneAssessment = "Progressing"
+	// Completed means the control plane successfully completed updating and no problems are detected
+	ControlPlaneAssessmentCompleted ControlPlaneAssessment = "Completed"
+	// Degraded means the process of updating the control plane suffers from an observed problem
+	ControlPlaneAssessmentDegraded ControlPlaneAssessment = "Degraded"
+)
+
+// ControlPlaneUpdateVersions contains the original and target versions of the upgrade
+type ControlPlaneUpdateVersions struct {
+	// previous is the version of the control plane before the update. When the cluster is being installed
+	// for the first time, the version will have a placeholder value '<none>' and carry 'Installation' metadata
+	// +required
+	// +kubebuilder:validation:XValidation:rule="self.version == '<none>' ? (has(self.metadata) && self.metadata.exists(m, m.key == 'Installation')) : !(has(self.metadata) && self.metadata.exists(m, m.key == 'Installation'))",message="previous version must be '<none>' iff marked with Installation metadata"
+	Previous Version `json:"previous"`
+
+	// target is the version of the control plane after the update. It may never be '<none>' or have `Installation` metadata
+	// +required
+	// +kubebuilder:validation:XValidation:rule="self.version != '<none>' && !(has(self.metadata) && self.metadata.exists(m, m.key == 'Installation'))",message="target version must not be '<none>' or have Installation metadata"
+	Target Version `json:"target"`
+}
+
+// Version describes a version involved in an update, typically on one side of an update edge
+type Version struct {
+	// version is a semantic version string, or a placeholder '<none>' for the special case where this
+	// is a "previous" version in a new installation, in which case the metadata must contain an item
+	// with key 'Installation'
+	// +required
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:MinLength=5
+	// +kubebuilder:validation:MaxLength=64
+	Version string `json:"version"`
+
+	// metadata is a list of metadata associated with the version. It is a list of key-value pairs. The value is optional
+	// and when not provided, the metadata item has boolean semantics (presence indicates true)
+	// +listType=map
+	// +listMapKey=key
+	// +patchStrategy=merge
+	// +patchMergeKey=key
+	// +optional
+	// +kubebuilder:validation:MaxItems=5
+	Metadata []VersionMetadata `json:"metadata,omitempty" patchStrategy:"merge" patchMergeKey:"key"`
+}
+
+// VersionMetadata is a key:value item assigned to version involved in the update. Value can be empty, then the metadata
+// have boolean semantics (true when present, false when absent)
+type VersionMetadata struct {
+	// key is the name of this metadata value
+	// +required
+	Key VersionMetadataKey `json:"key"`
+
+	// value is the value for the metadata
```

Comment:
What might the value actually contain?

Reply:
well, any short string. it's similar to labels conceptually.

Reply:
Explaining this in the godoc will be important prior to merging the API